crates/turborepo-filewatch/src/cookie_jar.rs

```diff
@@ -0,0 +1,341 @@
+            event = file_events.recv() => {
+                match unwrap_event(event) {
+                    Ok(event) => {
+                        if event.kind == EventKind::Create(notify::event::CreateKind::File) {
+                            let mut watches = watches.lock().await;
+                            for path in event.paths {
+                                let abs_path: &AbsoluteSystemPath = path
+                                    .as_path()
+                                    .try_into()
+                                    .expect("Non-absolute path from filewatching");
+                                if root.relation_to_path(abs_path) == PathRelation::Parent {
+                                    if let Some(responder) = watches.cookies.remove(&path) {
+                                        if let Err(_) = responder.send(Ok(())) {
+                                            // Note that cookie waiters will time out if they don't get a
+                                            // response, so we don't necessarily
+                                            // need to panic here, although we could decide to do that in the
+                                            // future.
+                                            debug!("failed to notify cookie waiter of cookie success");
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                    }
+                    Err(e) => {
+                        // we got an error, notify all waiters that their cookie failed
+                        let is_closing = matches!(
+                            e,
+                            WatchError::RecvError(broadcast::error::RecvError::Closed)
+                        );
+                        let resp = if is_closing { WatchError::Closed } else { e };
+                        let mut watches = watches.lock().await;
+                        for (_, sender) in watches.cookies.drain() {
+                            if let Err(_) = sender.send(Err(resp.clone())) {
+                                // Note that cookie waiters will time out if they don't get a response, so
+                                // we don't necessarily need to panic here, although
+                                // we could decide to do that in the future.
+                                debug!("failed to notify cookie waiter of error: {}", resp);
+                            }
+                        }
+                        if is_closing {
+                            watches.closed = true;
+                            return;
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+// result flattening is an unstable feature, so add a manual helper to do so.
+// This version is for unwrapping events coming from filewatching
+fn unwrap_event(
```

Comment:
Nit: I think `flatten_event` would be a lot better naming wise. `unwrap` make me think it'll panic if we encounter something that's not an event

Reply:
`flatten_event` is definitely better, switched